streamalert/alert_merger/main.py

```diff
@@ -73,17 +77,32 @@ def __init__(self):
         self.alert_proc_timeout = int(os.environ['ALERT_PROCESSOR_TIMEOUT_SEC'])
         self.lambda_client = boto3.client('lambda')
 
-    def _get_alerts(self, rule_name):
-        """Build a list of Alert instances triggered from the given rule name."""
-        alerts = []
+        # FIXME (derek.wang) Maybe make this configurable in the future
+        self._alert_generator_limit = self.ALERT_GENERATOR_DEFAULT_LIMIT
+
+    def _alert_generator(self, rule_name):
+        """
+        Returns a generator that yields Alert instances triggered from the given rule name.
 
+        To limit memory consumption, the generator yields a maximum number of alerts, defined
+        by self._alert_generator_limit.
+        """
+        alert_count = 0
         for record in self.table.get_alert_records(rule_name, self.alert_proc_timeout):
```

Comment:
omit the `alert_count` and use something like the following:
```python
for idx, record in enumerate(self.table.get_alert_records(rule_name, self.alert_proc_timeout), start=1):
    ...
    if idx >= self. _alert_generator_limit:
        ...
```

Reply:
(I'll test this on my own but regardless—) 
Does `enumerate` preserve the generator? It won't materialize the generator values up-front like `list`, right?

Reply:
@Ryxias yep it will preserve the generator underneath

Reply:
@Ryxias 
```python
def gen():
    for i in range(10):
        print('yielding', i)
        yield i


def main():
    iter = enumerate(gen(), start=1)
    print('created enumerator', iter)
    for i, v in iter:
        print('index: {}; value: {}'.format(i, v))


if __name__ == '__main__':
    main()
```

Output:
```
created enumerator <enumerate object at 0x10ddd1e60>
yielding 0
index: 1; value: 0
yielding 1
index: 2; value: 1
yielding 2
index: 3; value: 2
yielding 3
index: 4; value: 3
yielding 4
index: 5; value: 4
yielding 5
index: 6; value: 5
yielding 6
index: 7; value: 6
yielding 7
index: 8; value: 7
yielding 8
index: 9; value: 8
yielding 9
index: 10; value: 9
```